### DISCUSSION TITLE
Migrate Visual Snapshot Tests to Percy

### DISCUSSION THREAD

**Discussion by Sherlouk** [SwiftPackageIndex/SwiftPackageIndex-Server#1355] [Ideas]
As discussed on Discord.

https://percy.io/

Work with Dave to set everything up.

**Comment by daveverwer** [comment]
A quick update on this. I heard back from BrowserStack support team this morning and they do not currently include a Percy plan as part of their open-source sponsored account. I've asked if they have plans to add it, but I think I'm speaking to the wrong person about it.

The idea is good, but we wouldn't be able to implement this on their free plan.

**Comment by Sherlouk** [comment]
Alternatively: consider other third parties other than Percy who may have open-source sponsored accounts!

**Comment by daveverwer** [comment]
> Alternatively: consider other third parties other than Percy who may have open-source sponsored accounts!

Yes, of course. I had a look before posting this and couldn't see any that offer anything obvious. Are there any you'd recommend?

**Comment by Sherlouk** [comment]
I can't explicitly recommend any others as I've not tried them, but I have a few others which we've done some light investigations on so I'll do some digging and update you on my findings 👍

**Comment by finestructure** [comment]
Are we still pursuing this or should we close the issue?

**Comment by daveverwer** [comment]
I chased quite hard and got back two responses that said they are not currently supporting open-source projects with Percy and the cost of the product on its own is way out of our reach.

That said, it would be good to fix this problem. It's not really that great that all snapshot tests need to come through my machine.

**Comment by finestructure** [comment]
What we could do is set up a hosted runner using one of our Macs at MacStadium to run the snapshots and be the canonical source.

The only difficulty is to make sure whenever we change Xcode versions that machine can and does follow.

**Comment by daveverwer** [comment]
The other advantage of Percy was that it took the snapshots out of the repo. They are making it quite large already.

Given BrowserStack's track record with open-source, I'm sure they will add Percy to their open-source programme (which is very generous) at some point. Their [marketing page for the programme says](https://www.browserstack.com/open-source) that it covers "All products" and Percy is listed in the footer of that same page. 

I think we keep this open for a little while and I'll add a reminder to chase again in a month or so.

**Comment by Sherlouk** [comment]
Have we considered other third parties such as [Applitools](https://applitools.com/open-source/)? 
Should be same benefits, and has open-source pricing?

**Comment by finestructure** [comment]
I'm not super excited in general about tying this to a 3rd party service, to be honest. We don't have that many screens and the PR diffs have worked fine so far.

> The other advantage of Percy was that it took the snapshots out of the repo. They are making it quite large already.

Unfortunately, this would only affect future growth because the images are in the history already. (Also, at 18M `Tests` isn't super big I'd say.)
